File: Python/ex2_cashmachine.py
# 2. Create cashmachine that takes and withdraws only values of 50, 20, 10 and has a max account balance of 500

def digit(prompt, error):
    input_digit = input(prompt)
    while not input_digit.isdigit():
        input_digit = input(error)
    return int(input_digit)

class Account:
    def __init__(self):
        self.balance = 0

    def print(self):
        print("\nAccount balance is:", self.balance)
    
    def deposit(self):
        if self.balance >= 500:
            print("Cant deposit more money. Account balance cannot exceed 500 zł")
        else:    
            action = digit("How much money to deposit?: ", "Wrong amount of money. Enter a correct amount of money: ")
            if (action + self.balance) > 500:
                print("Cant deposit more money. Account balance cannot exceed 500 zł")
            else:
                if action % 50 == 0:
                    self.balance += action
                elif action % 20 == 0:
                    self.balance += action
                elif action % 10 == 0:
                    self.balance += action
                else:
                    print("Cashmachine deposits only bank notes of 50, 20 and 10 zł value. Enter a different value.") 
       
    def withdraw(self):
        if self.balance <= 10:
            print("Cant pay out more money. Account balance cannot be lower than 10 zł")
        else:
            action = digit("How much money to withdraw?: ", "Wrong amount of money. Enter a correct amount of money: ")
            if (self.balance - action) < 10:
                print("Cant pay out more money. Account balance be lower than 10 zł")
            else:
                if action % 50 == 0:  
                    self.balance -= action
                elif action % 20 == 0:
                    self.balance -= action
                elif action % 10 == 0:
                    self.balance -= action
                else:
                    print("Cashmachine pays out only bank notes of 50, 20 and 10 zł value. Enter a different value.")
    
    def exit(self):
        print("\nLogged out")
        print("Balance is:", self.balance, "zł")

if __name__ == "__main__":
    try:
        choice = 0
        account = Account()
        while account.balance < 10 or account.balance > 500:
            account.balance = digit("Enter an amount of account balance between 10 and 500 zł: ", "Wrong amount of money. Enter a correct amount of account balance between 10 and 500 zł: ")
            if account.balance < 10 or account.balance > 500:
                print("Wrong amount of account balance")
            else:
                while True:
                    print("\nMenu\n1 = Show account balance\n2 = Deposit money\n3 = Pay out money\n4 = Exit\n")
                    choice = digit("Choose an action: ", "Wrong action. Choose a correct action: ")
                    match choice:
                        case 1:
                            account.print()
                        case 2:
                            account.deposit()
                        case 3:
                            account.withdraw() 
                        case 4:
                            account.exit()
                            break
                        case _:
                            print("Wrong action")
    except Exception as error:
        print(f"Error!!, program shut down due to {error}")

File: JavaScript/ex2_cashmachine.js
// 2. Create cashmachine that withdraws only values of 50, 20 and 10 and has a max account balance of 500

const prompt = require ("prompt-sync")();

function digit(prompt_text, error){
    let input_digit = prompt(prompt_text);
    while (Number.isInteger(+input_digit) !== true){
        input_digit = prompt(error);
    }
    return parseInt(input_digit, 10);
}

class Account{
    constructor(){
        this.balance = 0;
    }

    print(){
        console.log("\nAccount balance:", this.balance)
    }

    deposit(){
        let action = 0;
        if (this.balance >= 500){
            console.log("Cant deposit more money. Account balance cannot exceed 500 zł")
        }
        else{
            action = digit("How much money to deposit?: ", "Wrong amount of money. Enter a correct amount of money: ");
            if ((action + this.balance) > 500){
                console.log("Account balance cannot exceed 500 zł");
            }
            else {
                if (action % 50 === 0){
                    this.balance += action;
                }
                else if (action % 20 === 0){
                    this.balance += action;
                }
                else if (action % 10 === 0 ){
                    this.balance += action;
                }
                else {
                    console.log("Cashmachine deposits only bank notes of 50, 20 and 10 zł value. Enter a different value.")
                }
            }
        }
    }

    withdraw(){
        let action = 0;
        if (this.balance <= 10){
            console.log("Cant pay out more money. Account balance cannot be lower than 10 zł")
        }
        else{
            action = digit("How much money to withdraw?: ", "Wrong amount of money. Enter a correct amount of money: ");
            if ((this.balance - action) < 10){
                console.log("Account balance be lower than 10 zł");
            }
            else {
                if (action % 50 === 0){
                    this.balance -= action;
                }
                else if (action % 20 === 0){
                    this.balance -= action;
                }
                else if (action % 10 === 0 ){
                    this.balance -= action;
                }
                else {
                    console.log("Cashmachine pays out only bank notes of 50, 20 and 10 zł value. Enter a different value.")
                }
            }
        }
    }

    exit(){
        console.log("\nLogged out")
        console.log("Balance is:", this.balance, "zł")
    }
}

if (require.main === module){
    try {
        const account = new Account();
        let choice = 0;
        while (account.balance < 10 || account.balance > 500){
            account.balance = digit("Enter an amount of account balance between 10 and 500 zł: ", "Wrong amount of money. Enter a correct amount of account balance between 10 and 500 zł: ");
            if (account.balance < 10 || account.balance > 500){
                console.log("Wrong amount of account balance");
            }
            else {
                while (true){
                    console.log("\nMenu");
                    console.log("1 = Show account balance:");
                    console.log("2 = Deposit money:");
                    console.log("3 = Pay out money:");
                    console.log("4 = Exit:");
                    choice = digit("Choose an action: ", "Wrong action. Choose a correct action: ");
                    switch (choice){
                        case (1):
                            account.print();
                            break;
                        case (2):
                            account.deposit();
                            break;
                        case (3):
                            account.withdraw();
                            break;
                        case (4):
                            account.exit();
                            return account.balance;
                        default:
                            console.log("Wrong action");
                            break;                    
                    }
                }
            }
        }
    }
    catch (error) {
        console.error(error)
    }
}